frontends/common/resolveReferences/resolveReferences.cpp

```diff
@@ -28,6 +28,43 @@ static const std::vector<const IR::IDeclaration *> empty;
 
 ResolutionContext::ResolutionContext() { anyOrder = P4CContext::get().options().isv1(); }
 
+Util::Enumerator<const IR::IDeclaration *> *ResolutionContext::getDeclarations(
+    const IR::INamespace *ns) const {
+    auto nsIt = namespaceDecls.find(ns);
+    auto *declVec = nsIt != namespaceDecls.end() ? nsIt->second : memoizeDeclarations(ns);
+    return Util::Enumerator<const IR::IDeclaration *>::createEnumerator(*declVec);
+}
+
+std::vector<const IR::IDeclaration *> *ResolutionContext::memoizeDeclarations(
+    const IR::INamespace *ns) const {
+    auto decls = ns->getDeclarations();
+    if (auto nest = ns->to<IR::INestedNamespace>()) {
+        // boost bug -- trying to iterate with an adaptor over an unnamed temp crashes
+        auto temp = nest->getNestedNamespaces();
+        for (auto nn : boost::adaptors::reverse(temp)) decls = nn->getDeclarations()->concat(decls);
+    }
+
+    return (namespaceDecls[ns] = decls->toVector());
+}
+
+Util::Enumerator<const IR::IDeclaration *> *ResolutionContext::getDeclsByName(
+    const IR::INamespace *ns, cstring name) const {
+    auto nsIt = declNames.find(ns);
+    std::function<const IR::IDeclaration *(
+        const std::pair<const cstring, const IR::IDeclaration *> &)>
+        second = [](const auto &entry) { return entry.second; };
```

Comment:
I don't see a reason to explicitly write the type of the function you can declare it as `auto second = ...`.

Reply:
This will not work, because `Enumerator::map(...)` expects a `std::function`, and not a lambda:
```
/local/kfcripps/repos/p4c/frontends/common/resolveReferences/resolveReferences.cpp: In member function 'Util::Enumerator<const IR::IDeclaration*>* P4::ResolutionContext::getDeclsByName(const IR::INamespace*, cstring) const':
/local/kfcripps/repos/p4c/frontends/common/resolveReferences/resolveReferences.cpp:57:70: error: no matching function for call to 'Util::Enumerator<std::pair<const cstring, const IR::IDeclaration*> >::map(P4::ResolutionContext::getDeclsByName(const IR::INamespace*, cstring) const::<lambda(const auto:1&)>&)'
   57 |         return Util::enumerate(range.first, range.second)->map(second);
      |                                                                      ^
In file included from /local/kfcripps/repos/p4c/ir/vector.h:22,
                 from /local/kfcripps/repos/p4c/ir/indexed_vector.h:23,
                 from /local/kfcripps/repos/p4c/build/ninja/release/ir/ir-generated.h:10,
                 from /local/kfcripps/repos/p4c/ir/ir.h:21,
                 from /local/kfcripps/repos/p4c/frontends/common/resolveReferences/resolveReferences.h:20,
                 from /local/kfcripps/repos/p4c/frontends/common/resolveReferences/resolveReferences.cpp:17:
/local/kfcripps/repos/p4c/lib/enumerator.h:110:20: note: candidate: 'template<class S> Util::Enumerator<S>* Util::Enumerator<T>::map(std::function<S(const T&)>) [with S = S; T = std::pair<const cstring, const IR::IDeclaration*>]'
  110 |     Enumerator<S> *map(std::function<S(const T &)> map);
      |                    ^~~
/local/kfcripps/repos/p4c/lib/enumerator.h:110:20: note:   template argument deduction/substitution failed:
/local/kfcripps/repos/p4c/frontends/common/resolveReferences/resolveReferences.cpp:57:70: note:   'P4::ResolutionContext::getDeclsByName(const IR::INamespace*, cstring) const::<lambda(const auto:1&)>' is not derived from 'std::function<S(const std::pair<const cstring, const IR::IDeclaration*>&)>'
   57 |         return Util::enumerate(range.first, range.second)->map(second);
      |                                                                      ^
/local/kfcripps/repos/p4c/frontends/common/resolveReferences/resolveReferences.cpp:63:66: error: no matching function for call to 'Util::Enumerator<std::pair<const cstring, const IR::IDeclaration*> >::map(P4::ResolutionContext::getDeclsByName(const IR::INamespace*, cstring) const::<lambda(const auto:1&)>&)'
   63 |     return Util::enumerate(range.first, range.second)->map(second);
      |                                                                  ^
In file included from /local/kfcripps/repos/p4c/ir/vector.h:22,
                 from /local/kfcripps/repos/p4c/ir/indexed_vector.h:23,
                 from /local/kfcripps/repos/p4c/build/ninja/release/ir/ir-generated.h:10,
                 from /local/kfcripps/repos/p4c/ir/ir.h:21,
                 from /local/kfcripps/repos/p4c/frontends/common/resolveReferences/resolveReferences.h:20,
                 from /local/kfcripps/repos/p4c/frontends/common/resolveReferences/resolveReferences.cpp:17:
/local/kfcripps/repos/p4c/lib/enumerator.h:110:20: note: candidate: 'template<class S> Util::Enumerator<S>* Util::Enumerator<T>::map(std::function<S(const T&)>) [with S = S; T = std::pair<const cstring, const IR::IDeclaration*>]'
  110 |     Enumerator<S> *map(std::function<S(const T &)> map);
      |                    ^~~
/local/kfcripps/repos/p4c/lib/enumerator.h:110:20: note:   template argument deduction/substitution failed:
/local/kfcripps/repos/p4c/frontends/common/resolveReferences/resolveReferences.cpp:63:66: note:   'P4::ResolutionContext::getDeclsByName(const IR::INamespace*, cstring) const::<lambda(const auto:1&)>' is not derived from 'std::function<S(const std::pair<const cstring, const IR::IDeclaration*>&)>'
   63 |     return Util::enumerate(range.first, range.second)->map(second);
      |                                                                  ^
[58/264] Building CXX object frontends/CMakeFiles/frontend.dir/p4/actionsInlining.cpp.o
ninja: build stopped: subcommand failed.
```

Reply:
> I don't see a reason to explicitly write the type of the function you can declare it as `auto second = ...`.

You cannot due to way stuff in `enumerators.h` is written. It cannot deduce type.

Reply:
> > I don't see a reason to explicitly write the type of the function you can declare it as `auto second = ...`.
> 
> You cannot due to way stuff in `enumerators.h` is written. It cannot deduce type.

You are right, I can see it now.